### ISSUE TITLE
How to return authorization header immediately after login with MSAL to Apollo-Graphql?

### ISSUE THREAD

**Issue by ivan7707** [AzureAD/microsoft-authentication-library-for-js#4449]
### Core Library

MSAL.js v2 (@azure/msal-browser)

### Core Library Version

2.18.0

### Wrapper Library

MSAL React (@azure/msal-react)

### Wrapper Library Version

1.1.0

### Description

Hello, I am using the standard apollo-client setup with the Micrososft authentication library MSAL.

Once someone is logged in, the authorization token is not set before the first render, so the useQuery in my app.js does not return data.

So upon login, the bearer token is set as undefined, however the JWT is in session storage. When immediately refreshing the page, the useQuery works, as it picks up the JWT.

How would I use MSAL in this case to return the JWT (Authorization header), immediately after the login process completes?

Thanks, 

### Relevant Code Snippets

```javascript
const httpLink = createHttpLink({
  uri: "http://localhost:5000/graphql",
});

const authLink = setContext(async (_, { headers }) => {
  // get the authentication token from local storage if it exists
  const token = await returnToken(
    msalInstance,
    protectedResources.apiHello.scopes
  );

  // return the headers to the context so httpLink can read them
  return {
    headers: {
      ...headers,
      authorization: `Bearer ${token}`,
    },
  };
});


const client = new ApolloClient({
  link: authLink.concat(httpLink),
  cache: new InMemoryCache(),
  connectToDevTools: true,
});


// MSAL function to get access token

async function returnToken(msalInstance: any, scope: any) {
  const accounts = msalInstance.getAllAccounts();
  const account = accounts[0];

  const accessTokenRequest = {
    scopes: scope,
    account: account,
  };

  if (account) {
    let resp = await msalInstance
      .acquireTokenSilent(accessTokenRequest)
      .then(function (accessTokenResponse: any) {
        // Acquire token silent success
        let accessToken = accessTokenResponse.accessToken;
        return accessToken;
      })
      .catch(function (error: any) {
        console.log(error);
      });
    return resp;
  }
}
```

### Identity Provider

Azure B2C Basic Policy

### Source

External (Customer)

**Comment by jasonnutter** [comment]
@ivan7707 The process of processing the authorization code and trading for tokens is asynchronous, and as such, applications need to wait until that process is completed. MSAL React handles this for you through hooks such as `useMsal`: https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-react/docs/getting-started.md#acquiring-an-access-token

Please take a look at that document and the [MSAL React samples](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/msal-react-samples) and let us know if you have further questions.

cc @tnorling

**Comment by ivan7707** [comment]
Thank you @jasonnutter.  I am aware of the asynchronous nature of the call.  The process I am trying to get my head around is how to factor that into the apollo-client update which happens in index.js.  

I'll keep digging, and take a look at those documents you sent and report back.

**Comment by ivan7707** [comment]
Hi @jasonnutter, thanks for the help.  

I was able to resolve the issue. The info from the first link you sent above was helpful, as it confirmed I was retrieving the access token correctly from the index.js page  **Acquiring an access token outside of a React component**.

**Solution:** 

The data query for the user data was being called immediately after the login redirect from the MSAL flow back to the main application prior to the tokens being set in SessionStorage.  I noticed that after login, there is a bit of latency until the AuthenticatedTemplate kicks in as well.  

In the react components that retrieve the user data, I have to ensure that the account is set prior to calling the graphql query for the data.  I did so with Apollo Graphql by using the useLazyQuery hook within useEffect, instead of the useQuery hook by itself outside of useEffect. 

```
const UserInfo: React.FC = () => {
  const { instance, accounts, inProgress } = useMsal();
  const account = useAccount(accounts[0] || {});

  const [loadingQuery, { loading, error, data }] = useLazyQuery(RETURN_USER);

  useEffect(() => {
    if (account) {
      loadingQuery();
    }
  }, [account, instance]);

...

```
